File: src/Poker/game/Deck.java
package Poker.game;

import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Deck
{
   private Integer[] deck = new Integer[DeckConstants.NUM_CARDS];
   private int cardIndex;
   
   public Deck()
   {
      for (int i = 0; i < DeckConstants.NUM_CARDS; i++)
      {
         deck[i] = i;
      }
   }
   
   public void shuffle()
   {
      List<Integer> cards = Arrays.asList(deck);
      
      cardIndex = 0;
      long lSeed = System.nanoTime() / 1000;
      Random random = new Random(lSeed);
      
      int remaining = DeckConstants.NUM_CARDS;
      for (int i = 0; i < DeckConstants.NUM_CARDS; i++)
      {
         int randomIndex = random.nextInt(remaining);
         deck[i] = cards.remove(randomIndex);
         remaining--;
      }
   }
   
   public Card dealCard()
   {
      if (cardIndex == DeckConstants.NUM_CARDS - 1)
      {
         return null;
      }
      
      return new Card(deck[cardIndex++]);
   }
}

File: server/game/deck.js
let Card = require('./card');

let Deck = (suits = 4, values = 13) =>
{
	const NUM_CARDS = suits * values;
	
	let cardIndex = 0;
	let cards = [];
	for (let i = 0; i < NUM_CARDS; i++)
	{
		cards[i] = i;
	}
	
	let self =
	{
		shuffle: () =>
		{
			cardIndex = 0;
			let shuffledCards = [];
			let remaining = NUM_CARDS;
			for (let i = 0; i < NUM_CARDS; i++)
			{
				let randomIndex = Math.floor(Math.random() * remaining);
				shuffledCards[i] = cards.splice(randomIndex, 1)[0];
				remaining--;
			}
			
			cards = shuffledCards;
		},
		deal: () =>
		{
			if (cardIndex == NUM_CARDS)
			{
				return null;
			}
			
			return Card(cards[cardIndex++]);
		},
		dealCards: (number) =>
		{
			let cards = [];
			while (number > 0)
			{
				let card = self.deal();
				if (!card)
				{
					return null;
				}
				cards.push(card);
				number--;
			}
			return cards;
		}
	}
	
	return self;
}

module.exports = Deck;
